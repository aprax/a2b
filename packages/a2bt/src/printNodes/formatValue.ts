/**
 * Color formats a string using ANSI escape codes.
 *
 * @param character The string to be color formatted.
 * @param ansiColorCode A value between 30-37 and 90-97 inclusive. See https://en.wikipedia.org/wiki/ANSI_escape_code#Colors.
 *
 * @return The color formatted string.
 */
export default (character: string, ansiColorCode?: number) =>
  ansiColorCode
    ? `\x01\x1b[${ansiColorCode}m${character.trim()}\x1b[39m\x02`
    : character.trim();
