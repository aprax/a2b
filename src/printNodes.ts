const printNodes = (numbers: number[]) => {
	if (numbers.length < 1) {
		return console.log('---No arguments passed---');
	}
	if (numbers.length === 1) {
		return console.log(numbers[0])
	}
	//const root = convertToTreeNodes([1, 2, 3, 4, 5])
	const height = Math.floor(Math.log2(numbers.length));
	const width = 2 * height;

	let branchHeight = 1; //This is the last child in the tree
	// Each iteration is the branches from the root and then the last child child
	for (let x = 1; x <= height; x++) {
		branchHeight += Math.pow(2, x) + 1;
	}
	const branchWidth = 2 * branchHeight - 1;

	const output = Array(branchHeight);

	for (let x = 0; x < output.length; x++) {
		output[x] = Array(branchWidth).fill(' ')
	}

	const colPos = Math.floor(branchWidth / 2);
	const branchCount = Math.pow(2, height);

	output[0][colPos] = numbers[0];

	for (let y = 1; y <= branchCount; y++) {
		output[y][colPos - y] = '/';
		output[y][colPos + y] = '\\';
	}

	const numbersIndexToPosition = Array(numbers.length);
	numbersIndexToPosition[0] = [0, colPos];
	for (let x = 1; x < numbers.length; x++) {
		const parentIndex = Math.floor((x - 1) / 2)
		const parentPos = numbersIndexToPosition[parentIndex]; //Verify

		const rowLevel = Math.floor(Math.log2(x + 1));

		const parentBranchCount = Math.pow(2, Math.floor(height - rowLevel + 1)); // Verify
		const isLeft = (x - 1) % 2 === 0;

		const rowPosition = parentPos[0] + parentBranchCount + 1;

		const colPosition = isLeft ? parentPos[1] - parentBranchCount - 1 : parentPos[1] + parentBranchCount + 1;

		output[rowPosition][colPosition] = numbers[x];
		numbersIndexToPosition[x] = [rowPosition, colPosition]

		if (2 * x + 1 < numbers.length) {
			const branchCount = Math.pow(2, height - rowLevel);
			for (let y = rowPosition + 1; y <= rowPosition + branchCount; y++) {
				output[y][colPosition - y + rowPosition] = '/';
				output[y][colPosition + y - rowPosition] = '\\';
			}
		}
	}
	for (const row of output) {
		console.log(row.join(''))
	}
	console.log('');
	return output;
}

export default printNodes;