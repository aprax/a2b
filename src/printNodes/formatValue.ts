/**
 * Color formats a string.
 *
 * @param {string} character
 * @param {number} ansiColorCode A value between 30-37 and 90-97 inclusive.
 *
 * @return {string}
 */
export default (character: string, ansiColorCode?: number) =>
  ansiColorCode
    ? `\x01\x1b[${ansiColorCode}m${character.trim()}\x1b[39m\x02`
    : character.trim();
