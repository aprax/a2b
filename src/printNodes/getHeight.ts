export default (
  index: number,
  branchCount: number,
  nodes: unknown[],
  side: "left" | "right"
) => {
  let height = 0;
  let step = 1;
  while (index < nodes.length) {
    height += branchCount / Math.pow(2, step++) + 1;
    index = index * 2 + (side === "left" ? 1 : 2);
  }
  return height;
};
