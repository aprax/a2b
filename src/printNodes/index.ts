const printNodes: (nodes: Object[]) => string[][] = (nodes) => {
  if (nodes.length <= 1) {
    return [[]];
  }
  let cellLength = Math.max(...nodes.map((node) => node.toString().length));
  if (cellLength % 2 === 0) {
    cellLength++;
  }

  const height = Math.floor(Math.log2(nodes.length)) + 1;

  let branchHeight = 1; // This is the last child in the tree
  // Each iteration is the branches from the root and then the last child child
  for (let x = 1; x <= height; x++) {
    branchHeight += Math.pow(2, x) + 1;
  }
  const branchWidth = 2 * branchHeight - 1;

  const output = Array<string[]>(branchHeight);

  for (let x = 0; x < output.length; x++) {
    output[x] = Array<string>(branchWidth).fill(
      " ".padEnd(Math.ceil(cellLength))
    );
  }

  const colPos = Math.floor(branchWidth / 2);
  const branchCount = Math.pow(2, height);

  const root = nodes[0] || 0;
  output[0]![colPos] = root
    .toString()
    .padStart(Math.ceil(cellLength / 2))
    .padEnd(cellLength);

  for (let y = 1; y <= branchCount; y++) {
    output[y]![colPos - y] = "/"
      .padStart(Math.ceil(cellLength / 2))
      .padEnd(cellLength);
    output[y]![colPos + y] = "\\"
      .padEnd(Math.ceil(cellLength / 2))
      .padStart(cellLength);
  }

  const numbersIndexToPosition = Array(nodes.length);
  numbersIndexToPosition[0] = [0, colPos];
  for (let x = 1; x < nodes.length; x++) {
    const parentIndex = Math.floor((x - 1) / 2);
    const parentPos = numbersIndexToPosition[parentIndex];

    const rowLevel = Math.floor(Math.log2(x + 1));

    const parentBranchCount = Math.pow(2, Math.floor(height - rowLevel + 1));
    const isLeft = (x - 1) % 2 === 0;

    const rowPosition = parentPos[0] + parentBranchCount + 1;

    const colPosition = isLeft
      ? parentPos[1] - parentBranchCount - 1
      : parentPos[1] + parentBranchCount + 1;

    const value = nodes[x] || 0;
    output[rowPosition]![colPosition] = isLeft
      ? value
          .toString()
          .padStart(Math.ceil(cellLength / 2))
          .padEnd(cellLength)
      : value
          .toString()
          .padEnd(Math.ceil(cellLength))
          .padStart(Math.ceil(cellLength));
    numbersIndexToPosition[x] = [rowPosition, colPosition];

    if (2 * x + 1 < nodes.length) {
      const branchCount = Math.pow(2, height - rowLevel);
      for (let y = rowPosition + 1; y <= rowPosition + branchCount; y++) {
        output[y]![colPosition - y + rowPosition] = "/"
          .padStart(Math.ceil(cellLength / 2))
          .padEnd(cellLength);
        output[y]![colPosition + y - rowPosition] = "\\"
          .padEnd(Math.ceil(cellLength / 2))
          .padStart(cellLength);
      }
    }
  }

  return output;
};

export default printNodes;
