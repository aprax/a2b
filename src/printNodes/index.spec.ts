import printNodes from ".";

const message = "should match the expected structure for ";
const MAX_LEVEL = 4; // Currently does not work for MAX_LEVEL >= 4

for (let x = MAX_LEVEL; x >= 0; x--) {
  const totalNodes = Math.pow(2, MAX_LEVEL - x) - 1;
  const numbers = Array(totalNodes);
  for (let y = 1; y <= totalNodes; y++) {
    numbers[y - 1] = y;
  }

  const actual = JSON.stringify(printNodes(numbers));
  it(`${message} ${JSON.stringify(numbers)}`, () => {
    expect(actual).toMatchSnapshot();
  });
}
