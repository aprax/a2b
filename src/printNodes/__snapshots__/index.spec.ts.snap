// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match the expected structure for  [] 1`] = `"[[]]"`;

exports[`should match the expected structure for  [1,2,3,4,5,6,7,8,9,10,11,12,13,14,15] 1`] = `"[[\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\" 1 \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\"],[\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\" / \\",\\"   \\",\\" \\\\\\\\ \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\"],[\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\" / \\",\\"   \\",\\"   \\",\\"   \\",\\" \\\\\\\\ \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\"],[\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\" / \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\" \\\\\\\\ \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\"],[\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\" / \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\" \\\\\\\\ \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\"],[\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\" / \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\" \\\\\\\\ \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\"],[\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\" / \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\" \\\\\\\\ \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\"],[\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\" / \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\" \\\\\\\\ \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\"],[\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\" / \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\" \\\\\\\\ \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\"],[\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\" 2 \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"3  \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\"],[\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\" / \\",\\"   \\",\\" \\\\\\\\ \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\" / \\",\\"   \\",\\" \\\\\\\\ \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\"],[\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\" / \\",\\"   \\",\\"   \\",\\"   \\",\\" \\\\\\\\ \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\" / \\",\\"   \\",\\"   \\",\\"   \\",\\" \\\\\\\\ \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\"],[\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\" / \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\" \\\\\\\\ \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\" / \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\" \\\\\\\\ \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\"],[\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\" / \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\" \\\\\\\\ \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\" / \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\" \\\\\\\\ \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\"],[\\"   \\",\\"   \\",\\"   \\",\\" 4 \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"5  \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\" 6 \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"7  \\",\\"   \\",\\"   \\",\\"   \\"],[\\"   \\",\\"   \\",\\" / \\",\\"   \\",\\" \\\\\\\\ \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\" / \\",\\"   \\",\\" \\\\\\\\ \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\" / \\",\\"   \\",\\" \\\\\\\\ \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\" / \\",\\"   \\",\\" \\\\\\\\ \\",\\"   \\",\\"   \\"],[\\"   \\",\\" / \\",\\"   \\",\\"   \\",\\"   \\",\\" \\\\\\\\ \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\" / \\",\\"   \\",\\"   \\",\\"   \\",\\" \\\\\\\\ \\",\\"   \\",\\"   \\",\\"   \\",\\" / \\",\\"   \\",\\"   \\",\\"   \\",\\" \\\\\\\\ \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\" / \\",\\"   \\",\\"   \\",\\"   \\",\\" \\\\\\\\ \\",\\"   \\"],[\\" 8 \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"9  \\",\\"   \\",\\"   \\",\\"   \\",\\"10 \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"11 \\",\\"   \\",\\"12 \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"13 \\",\\"   \\",\\"   \\",\\"   \\",\\"14 \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"   \\",\\"15 \\"]]"`;

exports[`should match the expected structure for  [1,2,3,4,5,6,7] 1`] = `"[[\\" \\",\\" \\",\\" \\",\\" \\",\\" \\",\\" \\",\\" \\",\\" \\",\\"1\\",\\" \\",\\" \\",\\" \\",\\" \\",\\" \\",\\" \\",\\" \\",\\" \\"],[\\" \\",\\" \\",\\" \\",\\" \\",\\" \\",\\" \\",\\" \\",\\"/\\",\\" \\",\\"\\\\\\\\\\",\\" \\",\\" \\",\\" \\",\\" \\",\\" \\",\\" \\",\\" \\"],[\\" \\",\\" \\",\\" \\",\\" \\",\\" \\",\\" \\",\\"/\\",\\" \\",\\" \\",\\" \\",\\"\\\\\\\\\\",\\" \\",\\" \\",\\" \\",\\" \\",\\" \\",\\" \\"],[\\" \\",\\" \\",\\" \\",\\" \\",\\" \\",\\"/\\",\\" \\",\\" \\",\\" \\",\\" \\",\\" \\",\\"\\\\\\\\\\",\\" \\",\\" \\",\\" \\",\\" \\",\\" \\"],[\\" \\",\\" \\",\\" \\",\\" \\",\\"/\\",\\" \\",\\" \\",\\" \\",\\" \\",\\" \\",\\" \\",\\" \\",\\"\\\\\\\\\\",\\" \\",\\" \\",\\" \\",\\" \\"],[\\" \\",\\" \\",\\" \\",\\"2\\",\\" \\",\\" \\",\\" \\",\\" \\",\\" \\",\\" \\",\\" \\",\\" \\",\\" \\",\\"3\\",\\" \\",\\" \\",\\" \\"],[\\" \\",\\" \\",\\"/\\",\\" \\",\\"\\\\\\\\\\",\\" \\",\\" \\",\\" \\",\\" \\",\\" \\",\\" \\",\\" \\",\\"/\\",\\" \\",\\"\\\\\\\\\\",\\" \\",\\" \\"],[\\" \\",\\"/\\",\\" \\",\\" \\",\\" \\",\\"\\\\\\\\\\",\\" \\",\\" \\",\\" \\",\\" \\",\\" \\",\\"/\\",\\" \\",\\" \\",\\" \\",\\"\\\\\\\\\\",\\" \\"],[\\"4\\",\\" \\",\\" \\",\\" \\",\\" \\",\\" \\",\\"5\\",\\" \\",\\" \\",\\" \\",\\"6\\",\\" \\",\\" \\",\\" \\",\\" \\",\\" \\",\\"7\\"]]"`;

exports[`should match the expected structure for  [1,2,3] 1`] = `"[[\\" \\",\\" \\",\\" \\",\\"1\\",\\" \\",\\" \\",\\" \\"],[\\" \\",\\" \\",\\"/\\",\\" \\",\\"\\\\\\\\\\",\\" \\",\\" \\"],[\\" \\",\\"/\\",\\" \\",\\" \\",\\" \\",\\"\\\\\\\\\\",\\" \\"],[\\"2\\",\\" \\",\\" \\",\\" \\",\\" \\",\\" \\",\\"3\\"]]"`;

exports[`should match the expected structure for  [1] 1`] = `"[[]]"`;
